File: backend/app/services/task_service.py
from sqlmodel import Session, select
from app.models.task import Task, TaskCreate, TaskUpdate
from app.models.user import User
from typing import List, Optional
from uuid import UUID


class TaskService:
    @classmethod
    def create_task(cls, db: Session, user_id: UUID, task_create: TaskCreate) -> Task:
        """
        Create a new task for a user
        """
        task = Task.from_orm(task_create)
        task.user_id = user_id
        
        db.add(task)
        db.commit()
        db.refresh(task)
        
        return task

    @classmethod
    def get_task_by_id(cls, db: Session, task_id: UUID, user_id: UUID) -> Optional[Task]:
        """
        Get a specific task by ID for a user
        """
        task = db.get(Task, task_id)
        if task and task.user_id == user_id:
            return task
        return None

    @classmethod
    def get_tasks_for_user(
        cls, 
        db: Session, 
        user_id: UUID, 
        limit: int = 50, 
        offset: int = 0,
        status_filter: Optional[str] = None
    ) -> List[Task]:
        """
        Get all tasks for a specific user with optional filtering and pagination
        """
        query = select(Task).where(Task.user_id == user_id)

        # Apply status filter if provided
        if status_filter:
            if status_filter == "active":
                query = query.where(Task.completed == False)
            elif status_filter == "completed":
                query = query.where(Task.completed == True)

        # Apply pagination
        query = query.offset(offset).limit(limit).order_by(Task.created_at.desc())

        return db.exec(query).all()

    @classmethod
    def update_task(cls, db: Session, task_id: UUID, user_id: UUID, task_update: TaskUpdate) -> Optional[Task]:
        """
        Update a task for a user
        """
        task = db.get(Task, task_id)
        if not task or task.user_id != user_id:
            return None

        # Update task with provided fields
        update_data = task_update.dict(exclude_unset=True)
        for field, value in update_data.items():
            setattr(task, field, value)

        db.add(task)
        db.commit()
        db.refresh(task)

        return task

    @classmethod
    def delete_task(cls, db: Session, task_id: UUID, user_id: UUID) -> bool:
        """
        Delete a task for a user
        """
        task = db.get(Task, task_id)
        if not task or task.user_id != user_id:
            return False

        db.delete(task)
        db.commit()
        return True

    @classmethod
    def toggle_task_completion(cls, db: Session, task_id: UUID, user_id: UUID) -> Optional[Task]:
        """
        Toggle the completion status of a task for a user
        """
        task = db.get(Task, task_id)
        if not task or task.user_id != user_id:
            return None

        task.completed = not task.completed
        db.add(task)
        db.commit()
        db.refresh(task)

        return task

File: frontend/src/services/taskService.ts
import { apiService } from './api';
import { Task, TaskFilterState } from '@/types';

interface CreateTaskData {
  title: string;
  description?: string;
  dueDate?: Date;
}

interface UpdateTaskData extends Partial<CreateTaskData> {
  completed?: boolean;
}

interface GetTasksParams {
  status?: 'all' | 'completed' | 'pending';
  page?: number;
  limit?: number;
}

class TaskService {
  /**
   * Get all tasks for the authenticated user
   */
  async getTasks(params?: GetTasksParams): Promise<Task[]> {
    try {
      const queryParams = new URLSearchParams();
      if (params?.status && params.status !== 'all') {
        queryParams.append('status', params.status);
      }
      if (params?.page) {
        queryParams.append('page', params.page.toString());
      }
      if (params?.limit) {
        queryParams.append('limit', params.limit.toString());
      }

      const queryString = queryParams.toString();
      const url = queryString ? `/tasks?${queryString}` : '/tasks';

      const response = await apiService.get<{ tasks: Task[] }>(url);
      return response.data.tasks.map(task => ({
        ...task,
        createdAt: new Date(task.createdAt),
        updatedAt: new Date(task.updatedAt),
        ...(task.dueDate && { dueDate: new Date(task.dueDate) })
      }));
    } catch (error: any) {
      throw new Error(error.response?.data?.error || 'Failed to fetch tasks');
    }
  }

  /**
   * Create a new task
   */
  async createTask(taskData: CreateTaskData): Promise<Task> {
    try {
      const response = await apiService.post<Task>('/tasks', {
        title: taskData.title,
        description: taskData.description,
        dueDate: taskData.dueDate ? new Date(taskData.dueDate).toISOString() : undefined
      });

      const newTask = response.data;
      return {
        ...newTask,
        createdAt: new Date(newTask.createdAt),
        updatedAt: new Date(newTask.updatedAt),
        ...(newTask.dueDate && { dueDate: new Date(newTask.dueDate) })
      };
    } catch (error: any) {
      throw new Error(error.response?.data?.error || 'Failed to create task');
    }
  }

  /**
   * Update an existing task
   */
  async updateTask(id: string, taskData: UpdateTaskData): Promise<Task> {
    try {
      const response = await apiService.put<Task>(`/tasks/${id}`, {
        ...taskData,
        ...(taskData.dueDate && { dueDate: new Date(taskData.dueDate).toISOString() })
      });

      const updatedTask = response.data;
      return {
        ...updatedTask,
        createdAt: new Date(updatedTask.createdAt),
        updatedAt: new Date(updatedTask.updatedAt),
        ...(updatedTask.dueDate && { dueDate: new Date(updatedTask.dueDate) })
      };
    } catch (error: any) {
      throw new Error(error.response?.data?.error || 'Failed to update task');
    }
  }

  /**
   * Toggle task completion status
   */
  async toggleTaskCompletion(id: string): Promise<Task> {
    try {
      const response = await apiService.patch<Task>(`/tasks/${id}/toggle-completion`);

      const updatedTask = response.data;
      return {
        ...updatedTask,
        createdAt: new Date(updatedTask.createdAt),
        updatedAt: new Date(updatedTask.updatedAt),
        ...(updatedTask.dueDate && { dueDate: new Date(updatedTask.dueDate) })
      };
    } catch (error: any) {
      throw new Error(error.response?.data?.error || 'Failed to toggle task completion');
    }
  }

  /**
   * Delete a task
   */
  async deleteTask(id: string): Promise<void> {
    try {
      await apiService.delete(`/tasks/${id}`);
    } catch (error: any) {
      throw new Error(error.response?.data?.error || 'Failed to delete task');
    }
  }

  /**
   * Filter tasks based on provided criteria
   */
  filterTasks(tasks: Task[], filters: TaskFilterState): Task[] {
    return tasks.filter(task => {
      // Filter by status
      if (filters.status === 'completed' && !task.completed) {
        return false;
      }
      if (filters.status === 'pending' && task.completed) {
        return false;
      }

      // Filter by search query
      if (filters.searchQuery) {
        const query = filters.searchQuery.toLowerCase();
        const matchesTitle = task.title.toLowerCase().includes(query);
        const matchesDescription = task.description?.toLowerCase().includes(query) || false;
        if (!matchesTitle && !matchesDescription) {
          return false;
        }
      }

      return true;
    }).sort((a, b) => {
      // Sort by specified field
      let comparison = 0;
      switch (filters.sortBy) {
        case 'title':
          comparison = a.title.localeCompare(b.title);
          break;
        case 'createdAt':
          comparison = a.createdAt.getTime() - b.createdAt.getTime();
          break;
        case 'updatedAt':
          comparison = a.updatedAt.getTime() - b.updatedAt.getTime();
          break;
        case 'dueDate':
          if (a.dueDate && b.dueDate) {
            comparison = a.dueDate.getTime() - b.dueDate.getTime();
          } else if (a.dueDate) {
            comparison = -1;
          } else if (b.dueDate) {
            comparison = 1;
          } else {
            comparison = 0;
          }
          break;
        default:
          comparison = 0;
      }

      // Apply sort direction
      return filters.sortDirection === 'asc' ? comparison : -comparison;
    });
  }
}

export const taskService = new TaskService();

export default TaskService;